### DISCUSSION TITLE
[V3] x-mask, updated hook and Form object

### DISCUSSION THREAD

**Discussion by ev3rlost** [livewire/livewire#8389] [1. Help]
So, I have the following input:

`<input wire:model="form.phone" x-mask="999 999 999" type="text" />`

And I've extracted a Form object for saving the data ...

Now I want to remove the whitespace which is added by Alpines x-mask before saving to the database.

This is my updated hook (inside the custom Form object):

```
    public function updatedPhone()
    {
        $this->phone = str_replace(' ', '', $this->phone);
    }
```

... but it doesn't seem to work. I still get the phone number with added whitespace.

I'm sure there's an easy way to this, but I'm kinda stuck and don't see the whole picture.

**Comment by EdgarsJoja** [comment]
afaik lifecycle hooks does not work outside components, you have to put this logic into the component, not form object.

EDIT: or just have some data format function called inside form object, just before saving/validating data.

**Comment by ev3rlost** [comment]
So I did the following, if anyone needs something like this in the future (no styling or extras for simplicity reasons):

I made a custom input blade component, just for phone numbers 
`<x-phone wire:model="phone" x-mask="999 999 999" />`

This is how the blade component looks like:
```
<div 
    x-data="{ masked: '', raw: '' }"
    x-effect="raw = masked.replace(/\D/g, '')"
>
    <input
        type="text"
        x-model="masked"
        {{ $attributes->whereStartsWith('x-mask') }}
    />
    <input
        type="hidden"
        x-modelable="raw"
        {{ $attributes->whereStartsWith('wire:model') }}
    >
</div>
```

So I ended up using two inputs, one is visible to the frontend using the `x-mask` directive for masking, the other hidden using the `wire:model` for data binding.

The way I synced both inputs is by using `x-effect` where I set the formated raw value to the hidden input - following by `x-modelable` to bind the value to the Livewire Form object.

If anyone has improvements, I'm all up for it!

**Reply by ev3rlost** [reply]
And If you want to retrieve the data (edit form for example) modify `x-data` as shown below:

```diff
-   x-data="{ masked: '', raw: '' }"
+   x-data="{ masked: $wire.$entangle('{{ $attributes->whereStartsWith('wire:model')->first() }}', raw: '' }"
```
